MapleServer2/PacketHandlers/Game/InteractObjectHandler.cs

```diff
@@ -61,9 +83,43 @@ private static void HandleUse(GameSession session, PacketReader packet)
                     session.Send(QuestPacket.CompleteQuest(item.Basic.QuestID));
                     break;
                 }
+                session.Send(InteractActorPacket.UseObject(actor));
+            }
+            if (actor.Type == InteractActorType.Gathering)
+            {
+                RecipeMetadata recipe = RecipeMetadataStorage.GetRecipe(actor.RecipeId);
+                long requireMastery = int.Parse(recipe.RequireMastery);
+                Enums.MasteryType type = (Enums.MasteryType) int.Parse(recipe.MasteryType);
+                long currentMastery = 0;
+                int stamina = 10;
+                MasteryExp mastery = session.Player.Levels.MasteryExp.FirstOrDefault(x => x.Type == (byte) type);
+                if (mastery != null)
+                {
+                    currentMastery = mastery.CurrentExp;
+                    stamina = mastery.Stamina;
+                }
+                List<RecipeItem> items = RecipeMetadataStorage.GetResult(recipe);
+                Random rand = new Random();
+                InteractObjecResult result = InteractObjecResult.Fail;
+                foreach (RecipeItem item in items)
+                {
+                    if ((currentMastery >= requireMastery)
```

Comment:
Doesn't seem like either of these change in the loop so this can be done outside the loop